Add REST client

patricklatimer opened the pull request:
Simple client object to interact with the dataverse REST API.

patricklatimer committed:
Add first stab at rest client

patricklatimer committed:
Update readme

patricklatimer committed:
Clean up rest client code

patricklatimer committed:
Remove example test file from template

patricklatimer committed:
Add test

patricklatimer commented:
Okay, this is more or less solid, though we may want to add a `delete_entry` method.

I was also playing with [odata query filters](https://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part1-protocol/odata-v4.0-errata03-os-part1-protocol-complete.html#_The_$filter_System) and may add another file listing out what those are so users don't have to know how to format their strings (something like `equal = lambda column, value: f"{column} eq {value}"`.

Alternatively, this [python-odata](https://github.com/eblis/python-odata) looks cool, and even generates a python file with the database schema. I'm having some credential trouble at the moment but I'll see if I can get this working as well.

patricklatimer committed:
Fix missing space in scope string

dyf commented on README.md:
```diff
@@ -8,6 +8,44 @@
+
+For instance, the dim_mice table has a dim_mice_baseid column with uids, but we'd like to refer to mice by their 6-7 digit mouse_id. This table should be given an Alternate Key for the mouse_id column, so you can refer to a mouse with `(mouse_id='614173')`
+
+## Repo setup
```
can you remove this bit from the readme?

dyf replied:
(the repo setup stuff)

dyf commented on src/dataverse_client/rest_client.py:
```diff
@@ -0,0 +1,205 @@
+    Loads from environment variables prefixed with "DATAVERSE_"
+    """
+
+    tenant_id: str = "32669cd6-737f-4b39-8bdd-d6951120d3fc"
```
can we load these defaults from a file instead?
e.g. JsonConfigSettings (https://docs.pydantic.dev/latest/concepts/pydantic_settings/#other-settings-source)

patricklatimer replied:
Do we want to commit to always having a canonical file to configure this library? For instance, we could deploy a config file to `C:/ProgramData/AllenInstitute/dataverse/dataverse_client_config.json` or something. We probably wouldn't want to have a password there though, so apps would still have to provide that.

If I were to use this in e.g. waterlog, I would probably put all this stuff in an entry in SIPE's keepass secrets database, then pull it from the application and provide it as kwargs to the `DataverseConfig` upon instantiation in the app.

dyf replied:
I think a canonical file for the defaults is fine? Then people can supply their own thing from wherever.

dyf commented on src/dataverse_client/rest_client.py:
```diff
@@ -0,0 +1,205 @@
+            client_credential=None,
+        )
+
+        if not self.config.username.endswith("@alleninstitute.org"):
```
if we're going to do this autocomplete (I don't think it's necessary, but I don't care that much), let's put this in a settings file somewhere.

patricklatimer replied:
Moved it to a computed_property in the config object.

patricklatimer committed:
remove repo setup readme section

patricklatimer committed:
Move "username@domain.org" check to config object

bruno-f-cruz reviewed (commented):
I am happy with the syntax of the library and I think we can start building around it.

However, (and maybe I missed it?) is there a way to have query language via their API or would we have to access via the sql one? Remember, in most cases, users will not have the uid of an entry and will have instead to query the table (eg: give me the most recent entry for animal foo)

patricklatimer commented:
> I am happy with the syntax of the library and I think we can start building around it.
> 
> However, (and maybe I missed it?) is there a way to have query language via their API or would we have to access via the sql one? Remember, in most cases, users will not have the uid of an entry and will have instead to query the table (eg: give me the most recent entry for animal foo)

Yeah, that would be cool. They have a whole odata query language where you can add do things like only request the top n entries, order_by a specific column, or filter by "column equals ..." or "column contains ...". It's not that complicated but it'd be nice to not require users to know it, so that's what I was saying about adding a file to encode what those look like.

For example, an odata_filters.py with things like:
```python
equal: Callable[[str, str], str] = lambda column, value: f"{column} eq {value}"
not_equal: Callable[[str, str], str] = lambda column, value: f"{column} ne {value}"
greater_than: Callable[[str, str], str] = lambda column, value: f"{column} gt {value}"
```

so you could do `client.query('my_table', options=[equal('column_name', 'value'])`.

I started to do that, then I found the [python-odata](https://github.com/eblis/python-odata) library that (if we want to use it) already has all those query options enumerated into a library, plus it autogenerates code for the table schema, and constructing queries looks pretty simple.

Anyway, yeah, query helpers for sure, maybe a different PR though because this technically works. I made an issue: #3

patricklatimer committed:
Add config yaml file in ProgramData

patricklatimer committed:
fix formatting

patricklatimer committed:
fix formatting

patricklatimer committed:
Don't create programdata directory on import

patricklatimer committed:
Update line-length to 100

patricklatimer committed:
Add odata queries

patricklatimer committed:
Add log messages for requests

patricklatimer commented:
Okay, new changes
- Load config values from a yaml file at C:/ProgramData/AllenInstitute/dataverse_client/config.yml
- Added a client.query method and Odata queries (filter, top, order_by, select)
- Added a log message like the below for each request

`INFO:dataverse_client.rest_client:Dataverse PATCH: "https://orgc1997c24.crm.dynamics.com/api/data/v9.2/crb81_dim_mice_bases(crb81_mouse_id='614174')", status code: 200, duration: 0.563041 seconds`

bruno-f-cruz commented on src/dataverse_client/rest_client.py:
```diff
@@ -0,0 +1,351 @@
+            f"duration: {response.elapsed.total_seconds()} seconds"
+        )
+
+        if not response.status_code == 200:
```
Consider using `request.raise_for_status()` instead of manually checking the 200 code

bruno-f-cruz commented on src/dataverse_client/rest_client.py:
```diff
@@ -0,0 +1,351 @@
+            ValueError: If the entry cannot be added.
+        """
+        url = self._construct_url(table)
+        response = requests.post(url, headers=self.headers, json=data)
```
Consider using the `timeout` argument in all requests to prevent calls from hanging indefinitely

patricklatimer committed:
Use Response.raise_for_status

patricklatimer committed:
Add request timeout

patricklatimer committed:
Remove unnecessary tests

patricklatimer committed:
Automatically instantiate config on client init

patricklatimer committed:
Update readme

bruno-f-cruz commented on src/dataverse_client/rest_client.py:
```diff
@@ -0,0 +1,345 @@
+            f"duration: {response.elapsed.total_seconds()} seconds"
+        )
+        response.raise_for_status()
+        return response.json()
```
There is an edge case here. Dataverse returns a 204 with no content for writes (at least the ones I was using). When you do reponse.json() this will throw a JSONDecodeError. Consider returning an Optional[Dict] or an empty Dict instead?

patricklatimer replied:
Oh, I fixed this for update_entry but must not have tested it here. It should fix it to just add  `headers = self.headers | {"Prefer": "return=representation"}` so it'll return the dict of the thing it just wrote

patricklatimer committed:
Update headers for add_entry to return the new entry

patricklatimer committed:
Move version to pyproject.toml
